### ISSUE TITLE
Interop+OpenSsl+SslException: SSL Read BIO failed with OpenSSL error

### ISSUE THREAD

**Issue by jchannon** [dotnet/corefx#10858]
I have an ASP.Net Core app running on OSX using Kestrel and HTTPS.

```
private void MyListener(CancellationToken token)
        {
            var certificate = new X509Certificate2(Path.Combine(Path.GetDirectoryName(typeof(MyClass).GetTypeInfo().Assembly.Location),"cert.pfx"),"");
            var urlPrefix = "https://*:";
            var url = urlPrefix + this.ListenerPort + "/";
            var host = new WebHostBuilder()
                .UseKestrel(options =>
                {
                    options.UseHttps(certificate);
                    options.NoDelay = true;
                    options.UseConnectionLogging();
                })
                .UseUrls(url)
                .UseStartup<MyStartup>()
                .Build();
           
            host.Run(token);
        }
```

The server starts fine but when a request is processed I get an exception logged.

```
warn: Microsoft.AspNetCore.Server.Kestrel[0]
      Connection processing ended abnormally
System.IO.IOException: The encryption operation failed, see inner exception. ---> Interop+OpenSsl+SslException: SSL Read BIO failed with OpenSSL error - .
   at Interop.OpenSsl.BioRead(SafeBioHandle bio, Byte[] buffer, Int32 count)
   at Interop.OpenSsl.Encrypt(SafeSslHandle context, Byte[] input, Int32 offset, Int32 count, Byte[]& output, SslErrorCode& errorCode)
   at System.Net.SslStreamPal.EncryptDecryptHelper(SafeDeleteContext securityContext, Byte[] input, Int32 offset, Int32 size, Boolean encrypt, Byte[]& output, Int32& resultSize)
   --- End of inner exception stack trace ---
   at System.Net.Security.SslStreamInternal.StartWriting(Byte[] buffer, Int32 offset, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslStreamInternal.ProcessWrite(Byte[] buffer, Int32 offset, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslStreamInternal.BeginWrite(Byte[] buffer, Int32 offset, Int32 count, AsyncCallback asyncCallback, Object asyncState)
   at System.Net.Security.SslStream.<>c.<WriteAsync>b__86_0(Byte[] bufferArg, Int32 offsetArg, Int32 sizeArg, AsyncCallback callback, Object state)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncImpl[TArg1,TArg2,TArg3](Func`6 beginMethod, Func`2 endFunction, Action`1 endAction, TArg1 arg1, TArg2 arg2, TArg3 arg3, Object state, TaskCreationOptions creationOptions)
   at System.Net.Security.SslStream.WriteAsync(Byte[] buffer, Int32 offset, Int32 size, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.Kestrel.Filter.Internal.LoggingStream.WriteAsync(Byte[] buffer, Int32 offset, Int32 count, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.Kestrel.Filter.Internal.StreamSocketOutput.WriteAsync(ArraySegment`1 buffer, Boolean chunk, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.Kestrel.Internal.Http.Frame.<ProduceEndAwaited>d__153.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Server.Kestrel.Internal.Http.Frame`1.<RequestProcessingAsync>d__2.MoveNext()
```

This is also an issue on Linux I believe after finding an issue on SO http://stackoverflow.com/questions/38960081/asp-net-core-linux-ssl-ssl-read-bio-failed-with-openssl-error

**Comment by jchannon** [comment]
I see this was added to 1.1.0 milestone. Any known workarounds?

**Comment by jchannon** [comment]
Pinging @joshfree @bartonjs 😄

**Comment by bartonjs** [comment]
Since it's only been reported today, no one has really had a chance to look at it.  So there's not really any known workaround, no.

**Comment by jchannon** [comment]
I tried the Kestrel sample and even that fails https://github.com/aspnet/KestrelHttpServer/blob/dev/samples/SampleApp/Startup.cs

Surely someone tested that on Win/Linux/OSX?

**Comment by jchannon** [comment]
Here's the interesting bit 
 ``` 
app.MapWhen(ctx => ctx.Request.Method == "POST",
               builder => builder.Run(async context =>
                     {
                         await context.Response.WriteAsync("hi");
                     
                     }
                 ));
```

no exception


 ```  
app.MapWhen(ctx => ctx.Request.Method == "POST",
             builder => builder.Run(context =>
                  {


                       context.Response.StatusCode = 201;
                       return Task.CompletedTask;

                   }
               ));
```
EXCEPTION

**Comment by bartonjs** [comment]
Hey, @stephentoub, you know async better than anyone else.  What's the difference in the success and failure cases here?

**Comment by stephentoub** [comment]
Can you point me to the code for builder.Run?  My guess is that it's taking a Func<Context, Task>, all of the cases are resulting in an exception, but in the async method case that exception is being caught and stored into the returned task, whereas in the non-async case the exception is escaping synchronously from the delegate invocation, and the builder.Run implementation is ignoring it or doing something else with it.

**Comment by andrewmcknight** [comment]
Not sure if it helps,  I also had this issue on linux only (worked fine on our  windows version)
(Note: we do not use web sockets so it's unrelated to that other issue ).

Turned out, the error occurred when the requests being made did not have any content in the response body. 
Some of ours are API functions, we send a POST request, and we only require HTTP status code, 200, 500 etc for the response. 
As soon as we added a plain text response e.g. "success" as well as the status codes , the error did not occur.

**Comment by jchannon** [comment]
Same as what I saw

**Comment by bartonjs** [comment]
This issue slipped through my cracks.  I'm trying to figure out what's going on here, but I've never used an ASP.Net project so I don't know how to set up the repro.  Does anyone have a minimal repro with idiot-proof setup instructions for me to achieve an investigation?

**Comment by jchannon** [comment]
Clone and run this

https://github.com/aspnet/KestrelHttpServer/tree/dev/samples/SampleApp

**Comment by anass-b** [comment]
@bartonjs I would like to mention as well that simply running the project in this repo: https://github.com/aspnet/KestrelHttpServer/tree/dev/samples/SampleApp may not make the bug appear.

Here is some steps that will help reproduce it:
- Use Linux or macOS (I could reproduce this issue on RHEL7)
- Setup your OS to trust the "testCert.pfx" certificate by adding it to the trusted list
- Have two ASP.NET Web Apps that talk to each other using REST (for example: WebApp1 will send some GET and POST requests to WebApp2)
- Both Web Apps should use HTTPS (you can duplicate the repo mentioned by @jchannon to create the WebApp1 and WebApp2 scenario)

**Comment by jchannon** [comment]
For me simply running that app in the repo on OSX showed the issue or
running the code I supplied on OSX

**Comment by halter73** [comment]
@jchannon I just ran the https://github.com/aspnet/KestrelHttpServer/blob/dev/samples/SampleApp/Startup.cs on debian 8.2 and everything seems to work.

First I tried making a request with Chrome ignoring cert errors which was successful. Then I generated a new non-expired testCert.pfx using the following openssl commands then added it as a trusted CA:

    debian82:~/KestrelHttpServer/samples/SampleApp$ openssl genrsa -out testCert.key 2048
    Generating RSA private key, 2048 bit long modulus
    .........................................................................................+++
    .............................................................+++
    e is 65537 (0x10001)
    debian82:~/KestrelHttpServer/samples/SampleApp$ openssl req -new -key testCert.key -out testRequest.csr
    You are about to be asked to enter information that will be incorporated
    into your certificate request.
    What you are about to enter is what is called a Distinguished Name or a DN.
    There are quite a few fields but you can leave some blank
    For some fields there will be a default value,
    If you enter '.', the field will be left blank.
    -----
    Country Name (2 letter code) [AU]:
    State or Province Name (full name) [Some-State]:
    Locality Name (eg, city) []:
    Organization Name (eg, company) [Internet Widgits Pty Ltd]:
    Organizational Unit Name (eg, section) []:
    Common Name (e.g. server FQDN or YOUR name) []:localhost
    Email Address []:

    Please enter the following 'extra' attributes
    to be sent with your certificate request
    A challenge password []:testPassword
    An optional company name []:
    debian82:~/KestrelHttpServer/samples/SampleApp$ openssl x509 -req -days 3650 -in testRequest.csr -signkey testCert.key -out testCert.crt
    Signature ok
    subject=/C=AU/ST=Some-State/O=Internet Widgits Pty Ltd/CN=localhost
    Getting Private key
    debian82:~/KestrelHttpServer/samples/SampleApp$ openssl pkcs12 -keypbe PBE-SHA1-3DES -certpbe PBE-SHA1-3DES -export -in testCert.crt -inkey testCert.key -out testCert.pfx
    Enter Export Password:
    Verifying - Enter Export Password:
    debian82:~/KestrelHttpServer/samples/SampleApp$ sudo cp testCert.crt /usr/local/share/ca-certificates/
    debian82:~/KestrelHttpServer/samples/SampleApp$ sudo vim /etc/ca-certificates.conf
    debian82:~/KestrelHttpServer/samples/SampleApp$ sudo update-ca-certificates -fresh

Next I made a request from another .NET Core app in case that was required for the repro:

```C#
﻿using System;
using System.Net.Http;

namespace ConsoleApplication
{
    public class Program
    {
        public static void Main(string[] args)
        {
            using (var client = new HttpClient())
            {
                var response = client.GetStringAsync("https://localhost:5001/").Result;
                Console.WriteLine("Response from https://localhost:5001/: {0}", response);
            }
        }
    }
}
```

It output `Response from https://localhost:5001/: hello, world` as expected.

**Comment by halter73** [comment]
I think I figured out what was going on. I finally started seeing the `Interop+OpenSsl+SslException: SSL Read BIO failed with OpenSSL error` when the app I was testing threw an exception prior to  writing to the response body.

A little further investigation using the LoggingConnectionFilter (which is enabled via the call to `options.UseConnectionLogging();` in the Kestrel SampleApp), showed that the SslException always followed a zero-length write. When I changed the app not to throw but to simply not write to the response body at all to verify the zero-length write theory, the SslException was still logged. This explains the behavior described in https://github.com/dotnet/corefx/issues/10858#issuecomment-240354193.

To further verify the zero length writes were the problem I changed the [LoggingStream](https://github.com/aspnet/KestrelHttpServer/blob/d64b4c7acb5840fbd279faa8362e1da84df8c16b/src/Microsoft.AspNetCore.Server.Kestrel/Filter/Internal/LoggingStream.cs#L103) used by the LoggingConnectionFilter to not pass writes with a count of 0 to the `_inner` SslStream. Sure enough, no more SslExceptions.
